cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationRunsStore.java

```diff
@@ -0,0 +1,384 @@
+    OperationRun currentRun = currentDetail.getRun();
+    OperationRun updatedRun = OperationRun.builder(currentRun)
+        .setStatus(OperationRunStatus.FAILED).setError(error).build();
+    OperationRunDetail<?> updatedDetail = OperationRunDetail.builder(currentDetail)
+        .setRun(updatedRun).setSourceId(sourceId).build();
+
+    Collection<Field<?>> fields = getCommonUpdateFields(runId);
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.DETAILS_FIELD,
+            GSON.toJson(updatedDetail)));
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.STATUS_FIELD,
+            GSON.toJson(OperationRunStatus.FAILED)));
+
+    StructuredTable operationRunsTable = getOperationRunsTable(context);
+    operationRunsTable.update(fields);
+  }
+
+  /**
+   * Get an operation run by id.
+   *
+   * @param runId {@link OperationRunId} for the run
+   * @throws OperationRunNotFoundException run with id does not exist in namespace
+   */
+  public OperationRunDetail<?> getOperation(OperationRunId runId)
+      throws NotFoundException, IOException {
+    Optional<StructuredRow> row = getOperationRunInternal(runId);
+    if (!row.isPresent()) {
+      throw new OperationRunNotFoundException(runId.getNamespace(), runId.getRun());
+    }
+    return GSON.fromJson(
+        row.get().getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+        OperationRunDetail.class
+    );
+  }
+
+  /**
+   * Scan operations in a namespace.
+   *
+   * @param request scan request including filters and limit
+   * @param txBatchSize batch size of transaction
+   */
+  public boolean scanRuns(ScanOperationRunsRequest request, int txBatchSize,
+      BiConsumer<String, OperationRunDetail<?>> consumer)
+      throws OperationRunNotFoundException, IOException {
+
+    AtomicReference<ScanOperationRunsRequest> requestRef = new AtomicReference<>(request);
+    AtomicReference<String> lastKey = new AtomicReference<>();
+    AtomicInteger currentLimit = new AtomicInteger(request.getLimit());
+
+    while (currentLimit.get() > 0) {
+      AtomicInteger count = new AtomicInteger();
+
+      scanRunsInternal(requestRef.get(), entry -> {
+        lastKey.set(entry.getKey());
+        currentLimit.decrementAndGet();
+        consumer.accept(entry.getKey(), entry.getValue());
+        return count.incrementAndGet() < txBatchSize && currentLimit.get() > 0;
+      });
+
+      if (lastKey.get() == null) {
+        break;
+      }
+
+      ScanOperationRunsRequest nextBatchRequest = ScanOperationRunsRequest
+          .builder(requestRef.get())
+          .setScanFrom(lastKey.get())
+          .setLimit(currentLimit.get())
+          .build();
+      requestRef.set(nextBatchRequest);
+      lastKey.set(null);
+    }
+    return currentLimit.get() == 0;
+  }
+
+
+  private void scanRunsInternal(ScanOperationRunsRequest request,
+      Function<Map.Entry<String, OperationRunDetail<?>>, Boolean> func)
```

Comment:
this doesn't need to be for a Map.Entry, the Operation detail should have the id.

Reply:
ok I think this is being used by the caller to keep track of the last key. It seems simpler if this method were to return the last key instead.

Reply:
updated